File: 2-Regression/4-Simple Linear Regression/simple_linear_regression_grade_predictor.R
# Simple Linear Regression

# STEP 1

# Importing the dataset
gradesdataset = read.csv('Grades_Data.csv')

# Splitting the dataset into the Training set and Test set
# install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(gradesdataset$Grade, SplitRatio = 2/3) # 2/3 in Training Set
training_set = subset(gradesdataset, split == TRUE)
test_set = subset(gradesdataset, split == FALSE)

# STEP 2

# Fitting Simple Linear Regression to the Training set
regressor = lm(formula = Grade ~ MinutesStudy, # f1 calls help for lm function
               data = training_set)

# Type summary(regressor) in console and notice *** for strength of dependence
# Low P value indicates independent variable is highly significant

# STEP 3

# Predicting the Test set Results
y_pred = predict(regressor, newdata = test_set)
# After execution, type y_pred in console to see values

# STEP 4

# Visualising the Training set results
# install.packages('ggplot2') // Commented out aftr install

# Call library
# library(ggplot2)

# Use functions to plot observation points, regression line, label axis. Points plotted with real observation data, but line with y predicted values
ggplot() +
  geom_point(aes(x = training_set$MinutesStudy, y = training_set$Grade),
             colour = 'red') +
  geom_line(aes(x = training_set$MinutesStudy, y = predict(regressor, newdata = training_set)),
            colour = 'blue') +
  ggtitle('Grade vs Minutes Studied (Training set)') +
  xlab('Minutes Studied') +
  ylab('Grade')


# Visualising the Test set results. Keep regression line oon training set, because it's a already trained on training set, using test set would obtain same line with new points.
ggplot() +
  geom_point(aes(x = test_set$MinutesStudy, y = test_set$Grade),
             colour = 'red') +
  geom_line(aes(x = training_set$MinutesStudy, y = predict(regressor, newdata = training_set)),
            colour = 'blue') +
  ggtitle('Grade vs Minutes Studied (Test set)') +
  xlab('Minutes Studied') +
  ylab('Grade')

File: 2-Regression/4-Simple Linear Regression/simple_linear_regression_grade_predictor.py
# Simple Linear Regression

# STEP 1

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
gradesdata = pd.read_csv('Grades_Data.csv')
X = gradesdata.iloc[:, :-1].values # -1 removes last col, makes matrix of features / independent variables
y = gradesdata.iloc[:, 1].values # 1 selects only 2nd col for dependent variable vector

# Splitting the dataset into the Training set and Test set, make test size a third 
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 1/3, random_state = 0)

# STEP 2

# Fitting Simple Linear Regression to the Training set
from sklearn.linear_model import LinearRegression # Import specific libray to make the regressor object of LinearRegression class
regressor = LinearRegression() # No params needed
regressor.fit(X_train, y_train) # Method that fits regressor object to the training sets, regressor is machine that learns on training set, can use this to make predictions on test set, regressor has learned correllations.

# STEP 3

# Predicting the Test set results
# Put predictions in vector, a vector of grade predictions based on test set
y_pred = regressor.predict(X_test) # Use predict method on test set

# compare y_test (actual salaries) with y_pred

# STEP 4

# Visualising the Training set results
plt.scatter(X_train, y_train, color = 'red') # Actual data (observation points) of training set plotted using scatter function
plt.plot(X_train, regressor.predict(X_train), color = 'blue') # Plot of predicted values on training set, regression line using plot function
# Notice above we plot predicted values of training set here
# Real values will bue red points, The blue line is the prediction regressor
plt.title('Grades vs Minutes of Study (Training set)')
plt.xlabel('Minutes of Study')
plt.ylabel('Grade')
plt.show()

# Visualising the Test set results (against same unique regression line built on training set)
plt.scatter(X_test, y_test, color = 'red')
plt.plot(X_train, regressor.predict(X_train), color = 'blue')
# Note we keep the train set for the regressor line because we use the training set for the maching to learn the correlations
plt.title('Grades vs Minutes of Study (Test set)')
plt.xlabel('Minutes of Study')
plt.ylabel('Grade')
plt.show()